src/index.js

```diff
@@ -183,20 +183,56 @@ export default function (babel) {
 
 
   // HELPER FUNCTIONS
+  function visitTailExpressions(path, allowLoops, visitor) {
+    path.resync();
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+    let tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (let tailPath of tailPaths) {
+      if (tailPath.isExpressionStatement()) {
+        visitor(tailPath);
+      } else if (tailPath.isFunctionDeclaration()) {
+        // Promote function declarations to exprs.
```

Comment:
Are we sure this is broken? Look at the output code. Remember that ultimately it's the javascript engine that runs it, so as long as the output code is correct, it will work.

I think a `FunctionExpression` with a name gets pretty printed by Babel into a declaration in the final code. That means that a modern JS engine will let you call it at a higher point in the scope because that's what ES5 says. So I actually think the output is correct.

One way or another, we definitely can't just blindly insert `return` because remember, these exprs are going to be `replacedWith` things like `array.push` or `obj[k]=v`. I'm not sure how those two things would square with each other. We would probably have to wrap in an IIFE.